Download button

kajarenc opened the pull request:
**st.download_button() widget implementation**

**Description:** This widget is a  button with only one predefined action: download provided file to the viewer's local computer. 

MediaFileManager is used as a file storage and st.button() widget as a prototype for the download button frontend part.
The widget should be stateless and raise exceptions when created inside a form.

kajarenc committed:
DownloadButton boilerplate code

kajarenc committed:
Make downloadButton works with text files

kajarenc committed:
add debug information in component
download button now accept bites and filenames as data

kajarenc committed:
download button can accept filenames, open files, string and bites data

kajarenc committed:
Make filename optional argument for download_button
Use generated hash with guessed extension if filename not provided

kajarenc committed:
Sent "Click" signal back to widget
Add types for Download button widget

kajarenc committed:
Sent "Click" signal back to widget
Add types for Download button widget

kajarenc committed:
fix comment

kajarenc committed:
Merge remote-tracking branch 'origin/download-button' into download-button

# Conflicts:
#	lib/streamlit/elements/button.py

kajarenc committed:
fix formatting

kajarenc committed:
Merge branch 'develop' into download-button

kajarenc committed:
* Raise Exception when download_button created inside st.form
* Remove ability to accept file name as a data parameter for Download Button

kajarenc committed:
Add download button to WidgetProto type

kajarenc committed:
Add dummy docstring

kajarenc committed:
fix typo

kajarenc committed:
Better error message for download_button invalid data

kajarenc committed:
Handle unicode file names correctly

kajarenc committed:
formatter fixes

kajarenc committed:
* Add session state rule check
* Remove unused HTML tag attributes
* Add TODO for missing docstring

kajarenc committed:
* TODOs added

kajarenc committed:
Formatter changes

kajarenc committed:
Download buttons with same file content and mimetype but different filenames now work correctly

kajarenc committed:
Add draft docstring for download_button widget

willhuang1997 commented on proto/streamlit/proto/Element.proto:
```diff
@@ -66,6 +67,7 @@ message Element {
     Balloons balloons = 12;
     BokehChart bokeh_chart = 17;
     Button button = 19;
+    DownloadButton download_button = 42;
```
Since we didn't coordinate on this, I think I put Metric as 42. Should I change my PR or just change it when we merge?

vdonato replied:
It's probably fine to just have whoever merges second have to change theirs (or flip a coin to see who has to do it)

vdonato commented on proto/streamlit/proto/DownloadButton.proto:
```diff
@@ -0,0 +1,27 @@
+message DownloadButton {
+  string id = 1;
+  string label = 2;
+  bool default = 3;
```
Huh, it's weird that this (and the Button) proto have default fields as I'm assuming they're both always `False`, but since there's a precedence set I'm indifferent to suggesting we get rid of this and thinking we should keep it.

willhuang1997 replied:
What does the default even mean in this proto?

vdonato commented on lib/streamlit/elements/button.py:
```diff
@@ -147,3 +237,41 @@ def deserialize_button(ui_value: bool, widget_id: str = "") -> bool:
+            is_for_static_download=True,
+        )
+        proto_download_button.url = this_file.url
+        return
```
nit: could this case be slightly rewritten so that it can use the code in lines 274 through 277 like the other cases?

willhuang1997 commented on lib/streamlit/elements/button.py:
```diff
@@ -147,3 +237,41 @@ def deserialize_button(ui_value: bool, widget_id: str = "") -> bool:
         return cast("streamlit.delta_generator.DeltaGenerator", self)
+
+
+def marshall_file(coordinates, data, proto_download_button, mimetype, filename=None):
```
super nit: filename should be file_name?

xaptronic replied:
filename, mimetype or file_name, mime_type - I feel like filename is pretty commonly one word..

willhuang1997 commented on lib/streamlit/media_file_manager.py:
```diff
@@ -44,7 +44,7 @@ def _get_session_id():
 
 
-def _calculate_file_id(data, mimetype):
+def _calculate_file_id(data, mimetype, filename=None):
```
same here

kajarenc committed:
rewrite if branches to reuse code

kajarenc committed:
Rename filename to file_name for download button changes

kajarenc committed:
Add first MediaFileManager unittests for download button

kajarenc committed:
Merge branch 'develop' into download-button

# Conflicts:
#	lib/streamlit/__init__.py

kajarenc committed:
Change download button id in Element.proto to resolve conflict with Metric widget

kajarenc committed:
Empty commit to rerun CI/CD

kajarenc committed:
Include `st.download_button` to exception text when try to be used with state

kajarenc committed:
draft version of generated filename, if filename not provided for download button

kajarenc committed:
empty commit to rerun CI

kajarenc committed:
empty commit to rerun CI

kajarenc committed:
move logic for generating filename from app title to separate function

kajarenc committed:
Add e2e and javascript unit tests boilerplate

kajarenc committed:
fix download file name generation error for download_button

kajarenc committed:
Merge branch 'develop' into download-button

kajarenc committed:
first download_button unit test

kajarenc committed:
add unittests

kajarenc committed:
Add snapshots test

kajarenc committed:
Remove default values for label and data in download_button

kajarenc committed:
Add snapshots for download button

kajarenc committed:
Add mark_for_delete field to solve issue with immediately deletion of files stored via downlaod_button under `if` statement in main script

kajarenc committed:
Formatting fix

kajarenc committed:
Remove wrong TODO

kajarenc committed:
Refactoring: Rename MediaFileManager class to InMemoryFileManager

kajarenc committed:
rename media_file_manager.py to in_memory_file_manager.py

kajarenc committed:
rename imports to in_memory_file_manager instead of media_file_manager

kajarenc committed:
Fix mypy error

kajarenc committed:
Remove ability to marshall numpy array in Download Button

kajarenc committed:
Empty commit to rerun CI

kajarenc committed:
Merge branch 'develop' into download-button

# Conflicts:
#	lib/streamlit/elements/button.py

vdonato committed:
Merge branch 'develop' into download-button

vdonato committed:
Add e2e test that st.download_button downloads file when clicked

After upgrading to Cypress 6.9.1, this was relatively trivial to do.

kajarenc committed:
remove print statement

kajarenc committed:
Merge remote-tracking branch 'origin/download-button' into download-button

kajarenc committed:
rename media_file_manager_test.py to in_memory_file_manager_test.py

kajarenc committed:
Refactoring: rename MeidaFile to InMemoryFile

kajarenc committed:
Remove file name from DownloadButton.proto (that information already saved corresponding in InMemoryFile)

kajarenc committed:
Download button key can be string or int
Documentation fixes

kajarenc committed:
refactor filename generation from page title
remove ` · Streamlit` suffix for autogenerated page title before creating filename from it

kajarenc committed:
Add type annotation to string_util functions

kajarenc committed:
Add unit tests for string_util functions

kajarenc committed:
fix broken test

kajarenc committed:
fix type in docstring

kajarenc committed:
Add unit tests for generate_download_filename_from_title

kajarenc committed:
Fix download button test

kajarenc committed:
Add type annotation for download_button data parameter

kajarenc committed:
Better variable names related to renaming MediaFile and MediaFileManager to InMemoryFile and InMemoryFileManager

kajarenc committed:
Empty commit to rerun CI

kajarenc committed:
Merge branch 'develop' into download-button

willhuang1997 commented on lib/streamlit/elements/button.py:
```diff
@@ -92,6 +98,89 @@ def button(
+        if help is not None:
+            download_button_proto.help = dedent(help)
+
+        def deserialize_button(ui_value, widget_id=""):
```
I see that this method is copied and pasted from the past button code. Is there a reason why this is declared here? Do we use this method in the download button API?

kajarenc replied:
This method is used in the download button`register_widget` call, and it is a way to receive value from the frontend back to the backend.

willhuang1997 commented on lib/streamlit/string_util.py:
```diff
@@ -55,3 +59,55 @@ def is_binary_string(inp):
+    underscores; and remove anything that is not an alphanumeric, dash,
+    underscore, or dot.
+    """
+    s = str(name).strip().replace(" ", "_")
```
Is there a reason to cast it as a string when type is already a string? I'm not super familiar with python but just wondering.

kajarenc replied:
Good question, @willhuang1997 
It is more or less can be avoided here, I think since we use this function only in one place with a string parameter.

But AFAIK specifying type doesn't guarantee you, that code will fail if the wrong type provided, since that type annotation doesn't work in runtime, if let's say user input data can be provided. 

I think it is okay, and a little bit safer to explicitly cast value here.

vdonato replied:
We can probably omit the cast in our codebase as I think mypy gives us the type safety to do so here, but I'd imagine it's written this way simply because that's how it was written in the django codebase this was taken from (see the link in the docstring)

vdonato commented on frontend/src/components/core/Block/Block.tsx:
```diff
@@ -542,6 +546,19 @@ class Block extends PureComponent<Props> {
       }
 
+      case "downloadButton": {
+        const downloadButtonProto = node.element
+          .downloadButton as DownloadButtonProto
```
nothing actionable here since the auto-formatter is probably what formatted things like this, but wow this formatting is weird

vdonato commented on frontend/src/components/widgets/DownloadButton/DownloadButton.test.tsx:
```diff
@@ -0,0 +1,97 @@
+  })
+
+  describe("UIButton props should work", () => {
+    it("onClick prop", () => {
+      const props = getProps()
+      const wrapper = shallow(<DownloadButton {...props} />)
+
+      const wrappedUIButton = wrapper.find(UIButton)
+
+      wrappedUIButton.simulate("click")
+
+      expect(
+        props.widgetMgr.setTriggerValue
+      ).toHaveBeenCalledWith(props.element, { fromUi: true })
+    })
```
It'd be nice if this test also verified that a link got clicked with an expected URL (can add this test in a followup PR since we're hoping to get this in soon and this is already covered by e2e tests).

kajarenc replied:
Added to my todo list :)

vdonato commented on lib/streamlit/elements/button.py:
```diff
@@ -92,6 +98,89 @@ def button(
+        key : str or int
+            An optional string or integer to use as the unique key for the widget.
+            If this is omitted, a key will be generated for the widget
+            based on its content. Multiple widgets of the same type may
+            not share the same key.
```
nit: I think this should now be "Multiple widgets may not share the same key". (And if it's not like this elsewhere, those docstrings should be updated too, but that's out of the scope of this PR)
